Docs: Blob transfer Architectural Decision Record (#233)
agera-edc/DataSpaceConnector: docs/developer/decision-records/2022-04-21-dpf-blob-data-transfer/README.md

3. Consumer creates a container where the Provider DPF may write blobs. The container is created only if the client specifies managedResources=true.
   The [ObjectStorageProvisioner](../../../../extensions/azure/blobstorage/blob-provision/src/main/java/org/eclipse/dataspaceconnector/provision/azure/blob/ObjectStorageProvisioner.java) is responsible for provisioning the container and for generating a SAS token to access the container.
   It creates the container and the SAS token by using the [BlobStoreApi](../../../../extensions/azure/blobstorage/blob-core/src/main/java/org/eclipse/dataspaceconnector/azure/blob/core/api/BlobStoreApi.java). [BlobStoreApiImpl](../../../../extensions/azure/blobstorage/blob-core/src/main/java/org/eclipse/dataspaceconnector/azure/blob/core/api/BlobStoreApiImpl.java) retrieves the storage account access key in the consumer vault. Then it can create the container and generate the SAS token.  
4. Consumer sends an IDS message to the Provider, containing the information needed to write data to the destination container. For example, the destination blob name and the SAS token needed to right blob to the container.  

> 5. Provider store the SAS token in its Vault.  

Reviewer: ```suggestion
5. Provider stores the SAS token in its Vault.  
```

6. Provider requests the blob transfer on the Provider DPF. The provider DPF can be embedded or run in a separated runtime. If it runs on a separated runtime, the Provider requests the transfer via an HTTP request.  
7. The Provider DPF gets the source storage account access key in the Provider Vault.  
8. The Provider DPF gets the SAS token needed to write the blob to the consumer blob container.  
9. The Provider DPF reads the data that needs to be transfered. The [AzureStorageDataSource](../../../../extensions/azure/data-plane/storage/src/main/java/org/eclipse/dataspaceconnector/azure/dataplane/azurestorage/pipeline/AzureStorageDataSource.java) provides the source data stream.  
10. The Provider DPF writes the data to the destination blob so that the consumer can access the data.
The [AzureStorageDataSink](../../../../extensions/azure/data-plane/storage/src/main/java/org/eclipse/dataspaceconnector/azure/dataplane/azurestorage/pipeline/AzureStorageDataSink.java) transfers the data to the blob destination.
When the transfer is finished, the Provider DPF needs to write a blob called `.complete`.  

> 11. In the meantime, the client polls the transfer status regularly on the consumer endpoint `/transferprocess/<PROCESS_ID>/state`. To determine if the transfer is completed, the consumer checks if a blob named `.complete` exists in the container.  

Reviewer: Would it make sense to add one more arrow between 11 and 12 with the call that is described in this point, that the consumer checks if a blob exists in the container (it would it be an arrow from Consumer to Consumer Storage Account)? It seems to be missing.

Author: Yes we can add it, I added it to the sequence diagram.

# Blob storage transfer

> ADR describing the blob storage transfer end to end flow between 2 participants.

Reviewer 2: Are those two participants what is called `consumer` and `provider` in the following text?

Author: The 2 participants would be 2 companies part of the Dataspace. During the data transfer, one EDC instance has the role of Consumer (consuming data), and one has the role of Provider. 
One EDC instance can be both Consumer and Provider, its role depends of the data transfer. I didn't add the explanation to the doc because I think this is a fair assumption that somebody reading this doc knows the concept of Consumer/Provider as this is a concept used in EDC.

## Description

The data-plane-azure-storage extension can be used on DPF to support blob transfers.
A client can trigger a blob transfer on the consumer side via the Data Management API.

> The client needs to use the managedResources=true option in its HTTP request, otherwise the transfer will not work. Client would not be aware that an error occurred, it would just never see the transferProcess with completed state when polling for the result.   

Reviewer 3: under what situation this will happen? Was not clear to me from the doc.

`Client would not be aware that an error occurred, it would just never see the transferProcess with completed state when polling for the result.`

Author: If any failure happen, the client/consumer cannot know and will just be waiting for the transfer to happen. There is an [issue](https://github.com/eclipse-dataspaceconnector/DataSpaceConnector/issues/1242) open for it.

> The client needs to use the managedResources=true option in its HTTP request, otherwise the transfer will not work. Client would not be aware that an error occurred, it would just never see the transferProcess with completed state when polling for the result.   

Reviewer 4: These are really two different issues.
- if using `managedResources=false`, the SAS URL is not generated, so the transfer will not work.
- the provider does not send to the consumer any message after the transfer was performed (succeeded or failed). The consumer only polls for the result. Therefore, if the transfer fails for *any* reason, the consumer is not aware of it and keeps polling indefinitely.

We should link to EDC issues for those/create them if necessary

Reviewer 4: Created
- https://github.com/eclipse-dataspaceconnector/DataSpaceConnector/issues/1242
- https://github.com/eclipse-dataspaceconnector/DataSpaceConnector/issues/1241

> The client needs to use the managedResources=true option in its HTTP request, otherwise the transfer will not work. Client would not be aware that an error occurred, it would just never see the transferProcess with completed state when polling for the result.   

Reviewer 5: nit: highlight the query string option
```suggestion
The client needs to use the `managedResources=true` option in its HTTP request, otherwise the transfer will not work. Client would not be aware that an error occurred, it would just never see the transferProcess with completed state when polling for the result.   
```

> Storage accounts access key should be stored in advanced in Keyvaults. The consumer can generate SAS token to give the provider the possibility to write data to its container.

Reviewer 3: nit: 
```suggestion
Storage accounts access key should be stored in Keyvaults before initiating transfers. The consumer can generate SAS token to give the provider the possibility to write data to its container.
```

## Sequence diagram

The following sequence diagram describes the flow to transfer a blob from a Provider storage account to a consumer storage account.

> It starts from the client triggering the transfer on the consumer side and finishes when the consumer deletes the blob

Reviewer 3: ```suggestion
It starts from the client triggering the transfer on the consumer side and finishes when the consumer triggers deprovision request for temporarily provisioned blob
```

> after the client triggered the data deletion.

Reviewer 3: ```suggestion
```

![blob-transfer](../../../diagrams/blob-transfer.png)

> 1. The client calls the data management API to trigger a transfer process. managedResources needs to be set to true, to make sure that the consumer provisions the blob container. If it is not set to true, the transfer will not work. Then, the transfer would never be completed. 

Reviewer 2: What is the use case for `POST /transferprocess` when `managedResources` is set to `false`? It would help to understand why it needs set to `true`.

Author: Makes sense thanks, I added an explanation.

2. Consumer gets the destination storage account access key in its Vault.  
3. Consumer creates a container where the Provider DPF may write blobs. The container is created only if the client specifies managedResources=true.
   The [ObjectStorageProvisioner](../../../../extensions/azure/blobstorage/blob-provision/src/main/java/org/eclipse/dataspaceconnector/provision/azure/blob/ObjectStorageProvisioner.java) is responsible for provisioning the container and for generating a SAS token to access the container.

>    It creates the container and the SAS token by using the [BlobStoreApi](../../../../extensions/azure/blobstorage/blob-core/src/main/java/org/eclipse/dataspaceconnector/azure/blob/core/api/BlobStoreApi.java). [BlobStoreApiImpl](../../../../extensions/azure/blobstorage/blob-core/src/main/java/org/eclipse/dataspaceconnector/azure/blob/core/api/BlobStoreApiImpl.java) retrieves the storage account access key in the consumer vault. Then it can create the container and generate the SAS token.  

Reviewer 3: nit: If you agree then maybe we can remove this part as it is not very clear.
```suggestion
```

Reviewer 4: Well I think we should document that generating a SAS token requires a storage account key, and the latter is retrieved from the key vault. Perhaps we could link to Azure doc for this

Author: @cpeeyush I removed the details about BlobStoreApiImpl class but I specified that the storage account key needs to be present in the Vault.

> 4. Consumer sends an IDS message to the Provider, containing the information needed to write data to the destination container. For example, the destination blob container name and the SAS token needed to right blob to the container.  

Reviewer 2: typo?

```suggestion
4. Consumer sends an IDS message to the Provider, containing the information needed to write data to the destination container. For example, the destination blob container name and the SAS token needed to write a blob to the container.  
```

> 5. Provider stores the SAS token in its Vault.  

Reviewer 2: Why is the SAS token stored in a vault and not directly included in step 6?

> 6. Provider requests the blob transfer on the Provider DPF. The provider DPF can be embedded or run in a separated runtime. If it runs on a separated runtime, the Provider requests the transfer via an HTTP request.  

Reviewer 2: typo & limiting to `HTTPS`

```suggestion
6. Provider requests the blob transfer on the Provider DPF. The provider DPF can be embedded or run in a separated runtime. If it runs on a separate runtime, the Provider requests the transfer via an HTTPS request.  
```

> 6. Provider requests the blob transfer on the Provider DPF. The provider DPF can be embedded or run in a separated runtime. If it runs on a separated runtime, the Provider requests the transfer via an HTTP request.  

Reviewer 3: nit: 
```suggestion
6. Provider initiates the blob transfer on the Provider DPF. The provider DPF can be embedded or run in a separated runtime. If it runs on a separated runtime, the Provider requests the transfer via an HTTP request.  
```

7. The Provider DPF gets the source storage account access key in the Provider Vault.  
8. The Provider DPF gets the SAS token needed to write the blob to the consumer blob container.  
9. The Provider DPF reads the data that needs to be transfered. The [AzureStorageDataSource](../../../../extensions/azure/data-plane/storage/src/main/java/org/eclipse/dataspaceconnector/azure/dataplane/azurestorage/pipeline/AzureStorageDataSource.java) provides the source data stream.  
10. The Provider DPF writes the data to the destination blob so that the consumer can access the data.
The [AzureStorageDataSink](../../../../extensions/azure/data-plane/storage/src/main/java/org/eclipse/dataspaceconnector/azure/dataplane/azurestorage/pipeline/AzureStorageDataSink.java) transfers the data to the blob destination.

> 11. When the transfer is finished, the Provider DPF writes a blob called `.complete`.

Reviewer 2: Should there a be clean up of the in step 5 stored SAS token in the vault?
If not, what component and process is responsible for it?

> 11. When the transfer is finished, the Provider DPF writes a blob called `.complete`.

Reviewer 2: Are there other states possible? E.g. something like `.failed` if the requested data is not available?

Author: For now there is no mechanism to make sure that the client or consumer knows if a transfer failed. There is an [issue](https://github.com/eclipse-dataspaceconnector/DataSpaceConnector/issues/1242) open for it.

> 12. In the meantime, the client polls the transfer status regularly on the consumer endpoint `/transferprocess/<PROCESS_ID>/state`.  

Reviewer 2: Is `PROCESS_ID` a value returned in step 1?

Author: Yes, I added the information in step 1 👍
